src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,445 @@
+            }
+            let height = (self.depth() - depth) as u32;
+            current_node.set_height(height);
+            self.storage
+                .insert(&current_node.hash(), current_node.as_buffer())?;
+        }
+
+        self.root_node = current_node;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::common::{Bytes32, StorageError, StorageMap};
+    use crate::sparse::{Buffer, MerkleTree};
+    use hex;
+
+    #[test]
+    fn test_root_returns_empty_sum_with_no_leaves() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let tree = MerkleTree::<StorageError>::new(&mut storage);
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_1() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..1 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "39f36a7cb4dfb1b46f03d044265df6a491dffc1034121bc1071a34ddce9bb14b";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_2() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..2 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "8d0ae412ca9ca0afcb3217af8bcd5a673e798bd6fd1dfacad17711e883f494cb";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_3() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..3 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "52295e42d8de2505fdc0cc825ff9fead419cbcf540d8b30c7c4b9c9b94c268b7";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_5() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..5 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "108f731f2414e33ae57e584dc26bd276db07874436b2264ca6e520c658185c6b";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
```

Comment:
Looks like we're missing "Test Root Update 10" from the spec

Reply:
Updated